### PR TITLE
WT-11878 __wt_cond_signal coredump bug fix

### PR DESCRIPTION
**Pull request by y123456yz**
In a multithreaded environment, we need to check whether cond has been released. For example, when __conn_close is run, the cond will all be released, but other threads may use the cond, which will coredump because the cond no longer exists.

### CODE REVIEW

**Comment by y123456yz** [comment]
line to your configuration file "/root/.gdbinit".
For more information about this security protection see the
"Auto-loading safe path" section in the GDB manual.  E.g., run from the shell:
        info "(gdb)Auto-loading safe path"
Core was generated by `./ex_hello -i 1'.
Program terminated with signal 11, Segmentation fault.
#0  __wt_cond_signal (session=0x26eda20, cond=0x0) at /root/yyz/mongodb-guanfang-master/master/wiredtiger-master/wiredtiger/src/os_posix/os_mtx_cond.c:168
168         if (cond->waiters == -1 || (cond->waiters == 0 && __wt_atomic_casi32(&cond->waiters, 0, -1)))
Missing separate debuginfos, use: debuginfo-install glibc-2.17-324.el7_9.x86_64
(gdb) bt
#0  __wt_cond_signal (session=0x26eda20, cond=0x0) at /root/yyz/mongodb-guanfang-master/master/wiredtiger-master/wiredtiger/src/os_posix/os_mtx_cond.c:168
#1  0x00007f52275b5e13 in __capacity_signal (session=<optimized out>) at /root/yyz/mongodb-guanfang-master/master/wiredtiger-master/wiredtiger/src/conn/conn_capacity.c:252
#2  0x00007f52275b660e in __wt_capacity_throttle (session=session@entry=0x26eda20, bytes=28563, type=WT_THROTTLE_CKPT)
    at /root/yyz/mongodb-guanfang-master/master/wiredtiger-master/wiredtiger/src/conn/conn_capacity.c:352
#3  0x00007f52274e58ac in __bm_write (bm=0x275a300, session=0x26eda20, buf=0x26e6e68, addr=0x7ffc19949dd0 "\207\207于\245\037[hnn\002", addr_sizep=0x7ffc19949ed8, 
    data_checksum=<optimized out>, checkpoint_io=true) at /root/yyz/mongodb-guanfang-master/master/wiredtiger-master/wiredtiger/src/block_cache/block_mgr.c:704
#4  0x00007f52274e4985 in __wt_blkcache_write (session=session@entry=0x26eda20, buf=buf@entry=0x26e6e68, addr=addr@entry=0x7ffc19949dd0 "\207\207于\245\037[hnn\002", 
    addr_sizep=addr_sizep@entry=0x7ffc19949ed8, compressed_sizep=compressed_sizep@entry=0x7ffc19949ed0, checkpoint=checkpoint@entry=false, checkpoint_io=true, 
    compressed=false) at /root/yyz/mongodb-guanfang-master/master/wiredtiger-master/wiredtiger/src/include/time_inline.h:111
#5  0x00007f52276e3443 in __rec_write (session=session@entry=0x26eda20, buf=buf@entry=0x26e6e68, addr=addr@entry=0x7ffc19949dd0 "\207\207于\245\037[hnn\002", 
    addr_sizep=addr_sizep@entry=0x7ffc19949ed8, compressed_sizep=compressed_sizep@entry=0x7ffc19949ed0, checkpoint=checkpoint@entry=false, checkpoint_io=true, 
    compressed=false) at /root/yyz/mongodb-guanfang-master/master/wiredtiger-master/wiredtiger/src/reconcile/rec_write.c:907
#6  0x00007f52276e6987 in __rec_split_write (session=session@entry=0x26eda20, r=r@entry=0x26e6c00, chunk=0x26e6d80, compressed_image=compressed_image@entry=0x0, 
    last_block=last_block@entry=false) at /root/yyz/mongodb-guanfang-master/master/wiredtiger-master/wiredtiger/src/reconcile/rec_write.c:2133
#7  0x00007f52276ec474 in __wt_rec_split (session=session@entry=0x26eda20, r=r@entry=0x26e6c00, next_len=232)
    at /root/yyz/mongodb-guanfang-master/master/wiredtiger-master/wiredtiger/src/reconcile/rec_write.c:1499
#8  0x00007f52276ec8f7 in __wt_rec_split_crossing_bnd (session=session@entry=0x26eda20, r=r@entry=0x26e6c00, next_len=<optimized out>)
    at /root/yyz/mongodb-guanfang-master/master/wiredtiger-master/wiredtiger/src/reconcile/rec_write.c:1604
#9  0x00007f52276d28d8 in __rec_row_leaf_insert (session=session@entry=0x26eda20, r=r@entry=0x26e6c00, ins=0x296ec30)
    at /root/yyz/mongodb-guanfang-master/master/wiredtiger-master/wiredtiger/src/reconcile/rec_row.c:614
#10 0x00007f52276d918b in __wt_rec_row_leaf (session=session@entry=0x26eda20, r=r@entry=0x26e6c00, pageref=pageref@entry=0x26be820, salvage=salvage@entry=0x0)
    at /root/yyz/mongodb-guanfang-master/master/wiredtiger-master/wiredtiger/src/reconcile/rec_row.c:1011
#11 0x00007f52276eae55 in __reconcile (session=session@entry=0x26eda20, ref=ref@entry=0x26be820, salvage=salvage@entry=0x0, flags=flags@entry=68, 
    page_lockedp=page_lockedp@entry=0x7ffc1994bb7f) at /root/yyz/mongodb-guanfang-master/master/wiredtiger-master/wiredtiger/src/reconcile/rec_write.c:291
#12 0x00007f52276ea0b5 in __wt_reconcile (session=session@entry=0x26eda20, ref=0x26be820, salvage=salvage@entry=0x0, flags=flags@entry=68)
    at /root/yyz/mongodb-guanfang-master/master/wiredtiger-master/wiredtiger/src/reconcile/rec_write.c:99
#13 0x00007f52275683a2 in __wt_sync_file (session=session@entry=0x26eda20, syncop=syncop@entry=WT_SYNC_CHECKPOINT)
    at /root/yyz/mongodb-guanfang-master/master/wiredtiger-master/wiredtiger/src/btree/bt_sync.c:442
#14 0x00007f522776cca5 in __checkpoint_tree (session=session@entry=0x26eda20, is_checkpoint=is_checkpoint@entry=true, cfg=cfg@entry=0x7ffc1994cb50)
    at /root/yyz/mongodb-guanfang-master/master/wiredtiger-master/wiredtiger/src/txn/txn_ckpt.c:2269
#15 0x00007f522776cfeb in __checkpoint_tree_helper (session=0x26eda20, cfg=0x7ffc1994cb50)

**Comment by y123456yz** [comment]
https://jira.mongodb.org/browse/WT-11878

**Comment by marcbutler** [comment]
Hi @y123456yz,
Would you be able to share the source to `ex_hello` where you saw the crash?
Thanks.

**Comment by y123456yz** [comment]
> Hi @y123456yz, Would you be able to share the source to `ex_hello` where you saw the crash? Thanks.

I've added some code logic to ex_hello, mainly:
1. Add io_capacity=(total=1M), which is about 10W data
2. for loop, write data
3. conn close


sorry， It's been so long, I don't have a record of the code.

**Comment by marcbutler** [comment]
Hi @y123456yz 
I have updated the issue. There is not sufficient information to investigate how this crash occurred without the `ex_hello` code (could the connection have been closed too early?) so I am closing the issue. Please re-open if you encounter it again.

**Comment by y123456yz** [comment]
> Hi @y123456yz I have updated the issue. There is not sufficient information to investigate how this crash occurred without the `ex_hello` code (could the connection have been closed too early?) so I am closing the issue. Please re-open if you encounter it again.

hi, @marcbutler 
The problem has recurred, the test code as following:

`        error_check(wiredtiger_open(home, NULL, "create, io_capacity=(total=1M),verbose=[all:1, metadata:0, api:0]", &conn));

        /* Open a session handle for the database. */
        error_check(conn->open_session(conn, NULL, NULL, &session));
        /*! [access example connection] */

        /*! [access example table create] */
        error_check(session->create(session, "table:access", "key_format=S,value_format=S"));
        /*! [access example table create] */

        /*! [access example cursor open] */
        error_check(session->open_cursor(session, "table:access", NULL, NULL, &cursor));
        /*! [access example cursor open] */

         //insert
        for (i = 0; i < MAX_TEST_KV_NUM; i++) {
            snprintf(buf, sizeof(buf), "key%d", i);
            cursor->set_key(cursor, buf);
			
            cursor->set_value(cursor, "old value  ###############################################################################################################################################################################################################\0");
            error_check(cursor->insert(cursor));
        }

        error_check(conn->close(conn, NULL)); /* Close all handles. */`

**Comment by y123456yz** [comment]
@marcbutler 
can you open this PR again, We need to follow up it, Because the problem hasn't been solved.

**Comment by marcbutler** [comment]
@y123456yz Reopened as requested. The example code you have copied here is that the entire contents of `main()`? I ask as I would like to get a reproduction to help understand why this is showing up in this particular circumstance. Thanks.

**Comment by y123456yz** [comment]
#include <test_util.h>

static const char *home = "WT_TEST";

int
main(void)
{
   // access_example(argc, argv);
   WT_CONNECTION *conn;
   WT_CURSOR *cursor;
   WT_SESSION *session;
//   const char  *value;
   int i;
   char buf[1024];

   
    error_check(wiredtiger_open(home, NULL, "create, io_capacity=(total=1M),verbose=[all:1, metadata:0, api:0]", &conn));
    /* Open a session handle for the database. */
    error_check(conn->open_session(conn, NULL, NULL, &session));
    /*! [access example connection] */

    /*! [access example table create] */
    error_check(session->create(session, "table:access", "key_format=S,value_format=S"));
    /*! [access example table create] */

    /*! [access example cursor open] */
    error_check(session->open_cursor(session, "table:access", NULL, NULL, &cursor));
    /*! [access example cursor open] */

     //insert
    for (i = 0; i < 100000; i++) {
        snprintf(buf, sizeof(buf), "key%d", i);
        cursor->set_key(cursor, buf);
		
        cursor->set_value(cursor, "old value  ###############################################################################################################################################################################################################\0");
        error_check(cursor->insert(cursor));
    }
    
    error_check(conn->close(conn, NULL)); /* Close all handles. */

    return (EXIT_SUCCESS);
}

**Comment by marcbutler** [comment]
@y123456yz 

Please find below the version of your test I am running. I found I needed to add `testutil_recreate_dir(home);` to get the test to work. I have also made some trivial changes, including removing the unnecessary string terminator`"\0"`.

I am seeing the crash: and I am investigating. At the moment I suspect the problem is that there is an attempt to throttle the IO during shutdown in the last checkpoint *after* the capacity service has been shutdown.

```c
#include <test_util.h>

static const char *home = "WT_TEST";

int main(void)
{
    WT_CONNECTION *conn;
    WT_CURSOR *cursor;
    WT_SESSION *session;
    int i;
    char buf[1024];

    testutil_recreate_dir(home);

    testutil_check(wiredtiger_open(home, NULL, "create, io_capacity=(total=1M),verbose=[all:1, metadata:0, api:0]", &conn));
    testutil_check(conn->open_session(conn, NULL, NULL, &session));

    testutil_check(session->create(session, "table:access", "key_format=S,value_format=S"));

    testutil_check(session->open_cursor(session, "table:access", NULL, NULL, &cursor));

    for (i = 0; i < 100000; i++) {
        snprintf(buf, sizeof(buf), "key%d", i);
        cursor->set_key(cursor, buf);

        cursor->set_value(cursor, "old value  ###############################################################################################################################################################################################################");
        testutil_check(cursor->insert(cursor));
    }

    testutil_check(cursor->close(cursor));
    testutil_check(conn->close(conn, NULL)); /* Close all handles. */

    return (EXIT_SUCCESS);
}
```

**Comment by y123456yz** [comment]
> ```c
> testutil_recreate_dir
> ```

@marcbutler yes, this is the all code, thanks.

**Comment by y123456yz** [comment]
@marcbutler 
I think the reason is as following:
the main thread free the cond(__conn_close->__wt_capacity_server_destroy) earlier than the eviction-server thread to evict the dirty  mem to disk(this will use __wt_capacity_throttle func to send signal __capacity_signal).
thanks.

**Comment by marcbutler** [comment]
@y123456yz Yes I agree that was my conclusion. Thank you for identifying this problem.

My initial reaction is that we should make sure the capacity server available until after the final checkpoint is complete, rather than prevent freeing the condition variable twice. However, that solution changes behavior: so it will need to be discussed within the team. I'm also curious why we call `__wt_capacity_server_destroy` twice during shutdown. I will post these questions on the associated Jira ticket to bring them to the attention of colleagues.

If you wanted to look at creating a fix that did just that please feel free to do so, but it still needs to be discussed so I don't know if it would be the accepted solution. Please also note I will be away for a week which will slow things down. Though colleage @korteland who is working on you with other tickets will still be available.

**Comment by marcbutler** [comment]
@y123456yz Can you please rebase on the latest develop to resolve the conflict. Thanks.

**Comment by y123456yz** [comment]
> @y123456yz Can you please rebase or merge from the latest develop to resolve the conflict. Thanks.

hi, @marcbutler , I have merge from develop and resolve the conflict, please review again.
thanks.

**Comment by marcbutler** [comment]
@y123456yz CI pipeline has been enabled.

File: src/conn/conn_capacity.c [hunk 1]
```diff
@@ -232,7 +232,8 @@ __capacity_signal(WT_SESSION_IMPL *session)
     conn = S2C(session);
     cap = &conn->capacity;
     if (cap->written >= cap->threshold && !cap->signalled) {
-        __wt_cond_signal(session, conn->capacity_cond);
+        if (conn->capacity_cond != NULL)
```
Comment (marcbutler) [inline src/conn/conn_capacity.c:235]: If the capacity server is being shutdown at the correct phase of teardown this null check is no longer required.

File: src/conn/conn_ckpt.c [hunk 2]
```diff
@@ -235,7 +235,7 @@ __wt_checkpoint_signal(WT_SESSION_IMPL *session, wt_off_t logsize)
 
     conn = S2C(session);
     WT_ASSERT(session, WT_CKPT_LOGSIZE(conn));
-    if (logsize >= conn->ckpt_logsize && !conn->ckpt_signalled) {
+    if (logsize >= conn->ckpt_logsize && !conn->ckpt_signalled && conn->ckpt_cond != NULL) {
```
Comment (marcbutler) [inline src/conn/conn_ckpt.c:238]: If the capacity server is being destroyed at the correct time during shutdown this null check is no longer required.

File: src/conn/conn_log.c [hunk 3]
```diff
@@ -647,7 +647,8 @@ __log_file_server(void *arg)
                 WT_ERR(__wt_close(session, &close_fh));
                 WT_ASSERT(session, __wt_log_cmp(&close_end_lsn, &log->sync_lsn) >= 0);
                 WT_ASSIGN_LSN(&log->sync_lsn, &close_end_lsn);
-                __wt_cond_signal(session, log->log_sync_cond);
+                if (log->log_sync_cond != NULL)
```
Comment (marcbutler) [inline src/conn/conn_log.c:650]: Another null check that can be removed.

**Comment by marcbutler** [comment]
@y123456yz Rather than go through and comment on every null check: these should not be required if the capacity server is being destroyed at the correct phase of shutdown and should be removed.

**Review by marcbutler — CHANGES_REQUESTED** [review]

File: src/conn/conn_compact.c [hunk 4]
```diff
@@ -729,7 +729,8 @@ __wt_background_compact_signal(WT_SESSION_IMPL *session, const char *config)
     conn->background_compact.config = stripped_config;
     stripped_config = NULL;
     conn->background_compact.signalled = true;
-    __wt_cond_signal(session, conn->background_compact.cond);
+    if (conn->background_compact.cond != NULL)
```
[the remark this answers is not in the document]
Comment (y123456yz) [reply src/conn/conn_compact.c]: hi, @marcbutler  
have removed the check for null.
thanks.

File: src/conn/conn_open.c [hunk 5]
```diff
@@ -102,6 +101,8 @@ __wt_connection_close(WT_CONNECTION_IMPL *conn)
 
     /* The eviction server is shut down last. */
     WT_TRET(__wt_evict_destroy(session));
+    /* The eviction server must behind __wt_evict_destroy, see WT-11878 */
```
Comment (ddanderson) [inline src/conn/conn_open.c:104]: The comment seems a bit unclear to me. Also, we don't generally reference tickets in comments unless they are FIXMEs.  I'd suggest:
```
/* The capacity server can only be shut down after all I/O is complete. */
```
Anyone that needs more detail can look at `git blame` and get to PRs and tickets.

**Review by ddanderson — APPROVED** [review]
The (non-comment) code changes look good.  I don't need to re-review, but the comment should be updated before merge.

**Comment by y123456yz** [comment]
@marcbutler @ddanderson 
I have change the comments, please review again, thanks.

**Review by marcbutler — APPROVED** [review]

**Pull request merged**